File: 2020/11/11101c/merge_sort.cpp
#include <cstring>
#include <iostream>

using namespace std;
int A[1000000];
int B[1000000];
void merge(int start, int end)
{
    int mid = (start + end) / 2;
    int left = start;
    int right = mid + 1;
    int i = 0;
    while (left <= mid && right <= end)
        B[i++] = A[left] <= A[right] ? A[left++] : A[right++];
    while (left <= mid)
        B[i++] = A[left++];
    while (right <= end)
        B[i++] = A[right++];
    for (int i = start; i <= end; i++)
        A[i] = B[i - start];
}

void mergesort(int start, int end)
{
    if (start == end)
        return;
    int mid = (start + end) / 2;
    mergesort(start, mid);
    mergesort(mid + 1, end);
    merge(start, end);
}
int main()
{
    freopen("merge_sort.txt", "r", stdin);
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int N;
    cin >> N;
    for (int i = 0; i < N; i++)
        cin >> A[i];
    mergesort(0, N - 1);
    for (int i = 0; i < N; i++)
        cout << A[i] << '\n';
    return 0;
}

File: BASE/JSBase/정렬/mergesort.js
const merge = (left, right) => {
  let arr = [];
  while (left.length && right.length) {
    if (left[0] < right[0]) {
      arr.push(left.shift());
    } else {
      arr.push(right.shift());
    }
  }
  return [...arr, ...left, ...right];
};

const mergesort = (nums) => {
  const mid = nums.length / 2;
  if (nums.length < 2) {
    return nums;
  }
  const left = nums.splice(0, mid);
  const right = nums;
  return merge(mergesort(left), mergesort(right));
};
console.log(mergesort([3, 1, 2, 5, 4]));
const arr1 = Array.from(Array(4), () => Array(4).fill(1));
console.log(arr1);
